### PR TITLE
[AR-5072] SDK fixes

### PR DESCRIPTION
**Pull request by makylfang**
# PR

## Describe your changes

- Added overlay for redirection if redirection flag is set
- Added preloading of iframe
- Updated `await auth.init()` to wait for connect if logged in

Screen Shot 2022-11-23 at 1 16 33 PM

## Issue ticket number and link

- [AR-5072](https://team-1624093970686.atlassian.net/browse/AR-5072)

## Checklist before requesting a review

- [x] You have performed a self-review of your own code
- [x] You are using approved terminology
- [x] Your changes have been tested locally
- [x] Your code builds clean without any errors or warnings

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -32,8 +33,26 @@ import { Chain } from './chainList'
...
 
+function preLoadIframe(url: string, appId: string) {
+  try {
+    if (typeof document === 'undefined') return
+    const iframeLink = document.createElement('link')
+    iframeLink.href = `${url}/${appId}/login`
+    iframeLink.type = 'text/html'
+    iframeLink.rel = 'prefetch'
+    if (iframeLink.relList && iframeLink.relList.supports) {
+      if (iframeLink.relList.supports('prefetch')) {
```
Comment (karthik-durai) [inline src/index.ts]: we can move the nested if condition to the first top on itself

`iframeLink.relList && iframeLink.relList.supports && iframeLink.relList.supports('prefetch')`

File: src/index.ts [hunk 2]
```diff
@@ -32,8 +33,26 @@ import { Chain } from './chainList'
...
+        document.head.appendChild(iframeLink)
+      }
+    }
+  } catch (error) {
+    console.warn(error)
+  }
+}
+
 class AuthProvider {
```
Comment (karthik-durai) [inline src/index.ts:37]: class name is `Arcana Provider` and importing a fn with a same name, can we rename any one?

cc: @ajithranka

File: src/styles.ts [hunk 3]
```diff
@@ -136,4 +136,39 @@ const widgetIframeStyle = {
...
+    height: '100%',
+    display: 'flex',
+    alignItems: 'center',
+    justifyContent: 'center',
+    flexDirection: 'column',
+    top: '0',
+    left: '0',
+    background: 'rgba(10, 10, 10)',
+    position: 'fixed',
+    zIndex: '100001',
```
Comment (karthik-durai) [inline src/styles.ts:151]: why 100001? for z-index?

File: src/utils.ts [hunk 4]
```diff
@@ -249,6 +255,70 @@ const getConstructorParams = (initParams?: Partial<ConstructorParams>) => {
   return p
 }
 
+const createOverlayOnRedirection = () => {
+  const shouldCreateOverlay = checkIfRedirection()
+  if (!shouldCreateOverlay) {
+    return
+  }
```
Comment (karthik-durai) [inline src/utils.ts]: `checkIfRedirection` can be called directly in the if.

example: `if(checkIfRedirection())`

File: src/utils.ts [hunk 5]
```diff
@@ -249,6 +255,70 @@ const getConstructorParams = (initParams?: Partial<ConstructorParams>) => {
...
+  overlay.appendChild(closeButton)
+
+  document.body.appendChild(overlay)
+}
+
+const checkIfRedirection = () => {
+  const redirectHash = window.location.hash
+  if (redirectHash) {
+    const searchParams = new URLSearchParams(redirectHash.substring(1))
+    const val = searchParams.get('fLR')
```
Comment (karthik-durai) [inline src/utils.ts:310]: what is `fLR`?

File: examples/index.js [hunk 6]
```diff
@@ -1,7 +1,21 @@
...
+  try {
+    console.time('auth_init')
+    await auth.init()
+    console.timeEnd('auth_init')
+    provider = auth.provider
+    console.log('Init auth complete!')
+    setHooks()
+  } catch (e) {
+    console.log({ e })
+    console.log(e)
```
Comment (ajithranka) [inline examples/index.js]:
```suggestion
    console.log({ e })
```

File: src/utils.ts [hunk 7]
```diff
@@ -249,6 +255,70 @@ const getConstructorParams = (initParams?: Partial<ConstructorParams>) => {
...
+    if (val === 'y') {
+      const cleanUrl = window.location.origin + window.location.pathname
+      window.history.replaceState(null, '', cleanUrl)
+      return true
+    }
+    return false
```
Comment (karthik-durai) [inline src/utils.ts]: I guess this `return` can be removed

File: examples/index.js (hunk 6 above)
Comment (karthik-durai) [inline examples/index.js]: we can remove the second `console.log`

File: src/index.ts (hunk 2 above)
Comment (makylfang) [reply src/index.ts:37]: @karthik-durai What?

File: src/styles.ts (hunk 3 above)
Comment (makylfang) [reply src/styles.ts:151]: its 1 greater than the widget zIndex so as to hide it.

File: src/utils.ts (hunk 5 above)
Comment (makylfang) [reply src/utils.ts:310]: its a key that is used by wallet-ui to mark it as redirection.

File: src/index.ts [hunk 8]
```diff
@@ -32,8 +33,26 @@ import { Chain } from './chainList'
...
 import { ModalController } from './ui/modalController'
 
+function preLoadIframe(url: string, appId: string) {
+  try {
+    if (typeof document === 'undefined') return
+    const iframeLink = document.createElement('link')
+    iframeLink.href = `${url}/${appId}/login`
+    iframeLink.type = 'text/html'
+    iframeLink.rel = 'prefetch'
+    if (iframeLink.relList && iframeLink.relList.supports) {
```
Comment (ajithranka) [inline src/index.ts]: This condition may be skipped. The [DOMTokenList API](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList#browser_compatibility) has full support across all browsers.

**Review by ajithranka — APPROVED** [review]

File: src/index.ts (hunk 1 above)
Comment (makylfang) [reply src/index.ts]: The whole condition is removed as per @ajithranka comment.

File: src/utils.ts (hunk 7 above)
Comment (makylfang) [reply src/utils.ts]: Removed.

File: examples/index.js (hunk 6 above)
Comment (makylfang) [reply examples/index.js]: Removed.

File: src/index.ts (hunk 8 above)
Comment (makylfang) [reply src/index.ts]: Removed.

File: src/index.ts (hunk 2 above)
Comment (karthik-durai) [reply src/index.ts:37]: sorry my bad, ignore this.

File: src/styles.ts (hunk 3 above)
Comment (karthik-durai) [reply src/styles.ts:151]: ok

File: src/utils.ts (hunk 5 above)
Comment (karthik-durai) [reply src/utils.ts:310]: ok

File: src/utils.ts (hunk 4 above)
Comment (makylfang) [reply src/utils.ts]: Changed name, directly being called now.

**Review by shrinathprabhu — APPROVED** [review]

**Comment by shaloo** [comment]
Instead of asking the user "continue on this window or use original window", why cant we make it simple for the user?  What will a user **continue** with on the current window anyway - it is blank. Am I missing something?

Just display a 'close' button in green right below login successful message. Let the user take note that login is successful and then click the close button to go back to the original window. No?

**Comment by makylfang** [comment]
> Instead of asking the user "continue on this window or use original window", why cant we make it simple for the user? What will a user **continue** with on the current window anyway - it is blank. Am I missing something?
> 
> Just display a 'close' button in green right below login successful message. Let the user take note that login is successful and then click the close button to go back to the original window. No?

@shaloo If you see the screenshot there is already a close button. Clicking that button can't take you back to the original window since this will appear only on passwordless link login, where there is no way to know the original window.
The reason this message appears is because user want to continue on the original window so as the flow isn't broken.

**Comment by shaloo** [comment]
> 

That is right.  We cant take the user back to the original window but the message is confusing at best. 

It seems to suggest 'either continue on this window or go back to the original'.  The path to continuing is to click close and go back to original (user does it on their own) - that is the second part of the OR clause.  So first part before OR does not mean much, right?

The message could simply say 'Click close and continue on the original window.' instead of the statement with the "OR" clause causing confusion.  The only meaningful action user can take on the current window is to close it and resume work on the original one.

ALSO The close button is too small and not well positioned.

**Pull request merged**